### PR TITLE
8236840: Memory leak when switching ButtonSkin

### PR DESCRIPTION
**Pull request by arapte**
ButtonSkin adds a `ChangeListener` to `Control.sceneProperty()` which results in leaking the `ButtonSkin` itself when the `Button`'s skin is changed to a new `ButtonSkin`. 
Using a `WeakChangeListener` instead of `ChangeListener` solves the issue.

Please take a look.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8236840](https://bugs.openjdk.java.net/browse/JDK-8236840): Memory leak when switching ButtonSkin

### Reviewers
 * Jeanette Winzenburg ([fastegal](@kleopatra) - Author)
 * Kevin Rushforth ([kcr](@kevinrushforth) - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/147/head:pull/147`
`$ git checkout pull/147`

### CODE REVIEW

**Comment by kevinrushforth** [comment]
In general, there are two approaches to avoiding listener-related memory leaks. One is to use a WeakListener; the other is to explicitly remove the listener when the object is removed or otherwise no longer needed.

Using a WeakListener is certainly easier, but runs the risk of the listener being removed too early and not cleaning up after itself. I'm not suggesting that's the case here, but it is worth looking at. The one thing I would ask you to take a look at is whether it would matter if the old skin didn't call `setDefaultButton(oldScene, false)` when removed (and similarly `setCancelButton`).

**Comment by kevinrushforth** [comment]
@aghaisas can you also review?

**Comment by arapte** [comment]
> 
> 
> Using a WeakListener is certainly easier, but runs the risk of the listener being removed too early and not cleaning up after itself. I'm not suggesting that's the case here, but it is worth looking at. 


The listener does not get early GCed here. I did verify this by creating large number of `ButtonSkin`s. The latest `ButtonSkin` set on the `Button` is never GCed

>
>
>The one thing I would ask you to take a look at is whether it would matter if the old skin didn't call `setDefaultButton(oldScene, false)` when removed (and similarly `setCancelButton`).
This seems to be a bigger issue.
If we create multiple `ButtonSkin`s for a given `Button`, then all the `ButtonSkin` register listeners with the properties of that `Button`. And it results in various listeners being added to same property of the `Button`. and this should be a common issue across all skins.

**Comment by arapte** [comment]
Hi Kevin, Please take a look at the updated changes.
Added few more tests as we discussed and an explicit call to remove the listener.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java [hunk 1]
```diff
@@ -78,6 +80,25 @@
...
+        if (newScene != null) {
+            if (getSkinnable().isDefaultButton()) {
+                setDefaultButton(newScene, true);
+            }
+            if (getSkinnable().isCancelButton()) {
+                setCancelButton(newScene, true);
+            }
+        }
+    };
+    WeakChangeListener<Scene> weakChangeListener = new WeakChangeListener<>(sceneChangeListener);
```
Comment (kleopatra) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java]: a loose naming convention for weak listeners seems to be to prepend _weak_ to the name of the strong listener, which would be 
    
    WeakChangeListener<Scene> weakSceneChangeListener = new WeakChangeListener<>(sceneChangeListener);

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java [hunk 2]
```diff
@@ -171,6 +175,7 @@ public ButtonSkin(Button control) {
 
     /** {@inheritDoc} */
     @Override public void dispose() {
+        getSkinnable().sceneProperty().removeListener(weakChangeListener);
```
Comment (kleopatra) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java]: +1! Actually, looks like that manual remove is really needed - otherwise we get an NPE when removing the default button from its parent after the skin has been switched: 

    @Test
    public void testDefaultButtonSwitchSkinAndRemove() {
        Button button = new Button();
        button.setDefaultButton(true);
        Group root = new Group(button);
        Scene scene = new Scene(root);
        Stage stage = new Stage();
        stage.setScene(scene);
        stage.show();
        
        button.setSkin(new ButtonSkin1(button));
        root.getChildren().remove(button);
    }

Note: to see this NPE as failing test (vs. its printout on sysout), we need to re-wire the uncaughtExceptionHandler, see ComboBoxTest setup/cleanup for an example.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java (hunk 1 above)
Comment (arapte) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java]: Corrected in the next commit...

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java (hunk 2 above)
Comment (arapte) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java]: Thanks for the test case, I did some minor changes to it and included in the next commit.
The NPE can occur even without `button.setDefaultButton(true);`.
Please take a look

Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java]: good catch! You are right, without explicit removal of the listener, the NPE happens always.

Actually, I had been sloppy and got distracted by the NPE from my actual goal which was to dig into Kevin's "not cleaning up after itself" and .. finally found a (concededly extreme corner-case :) where that's happening: when setting the skin to null. Two failing tests:

    @Test
    public void testDefaultButtonNullSkinReleased() {
        Button button = new Button();
        button.setDefaultButton(true);
        Group root = new Group(button);
        Scene scene = new Scene(root);
        Stage stage = new Stage();
        stage.setScene(scene);
        stage.show();
        WeakReference<ButtonSkin> defSkinRef = new WeakReference<>((ButtonSkin)button.getSkin());
        button.setSkin(null);
        
        attemptGC(defSkinRef);
        assertNull("skin must be gc'ed", defSkinRef.get());
    }
    
    @Test
    public void testDefaultButtonNullSkinAcceleratorRemoved() {
        Button button = new Button();
        button.setDefaultButton(true);
        Group root = new Group(button);
        Scene scene = new Scene(root);
        Stage stage = new Stage();
        stage.setScene(scene);
        stage.show();
        KeyCodeCombination key = new KeyCodeCombination(KeyCode.ENTER);

        assertNotNull(scene.getAccelerators().get(key));
        button.setSkin(null);
        assertNull(scene.getAccelerators().get(key));
        
    }

An explicitly cleanup in dispose makes them pass:

    @Override
    public void dispose() {
        setDefaultButton(false);
        setCancelButton(false);
        getSkinnable().sceneProperty().removeListener(weakChangeListener);

Comment (arapte) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/ButtonSkin.java]: I have included this change and the test, with slight modification to include same test for Cancel button.
